File: netlogo-gui/src/main/prim/etc/_anyturtleson.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.agent.AgentIterator;
import org.nlogo.core.AgentKindJ;
import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Patch;
import org.nlogo.agent.Turtle;
import org.nlogo.core.I18N;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.ArgumentTypeException;
import org.nlogo.nvm.RuntimePrimitiveException;
import org.nlogo.nvm.Reporter;

import java.util.ArrayList;
import java.util.List;

public final class _anyturtleson
    extends Reporter {

  @Override
  public Object report(final org.nlogo.nvm.Context context) throws LogoException {
    Object agentOrSet = args[0].report(context);
    if (agentOrSet instanceof Turtle) {
      Turtle turtle = (Turtle) agentOrSet;
      if (turtle.id() == -1) {
        throw new RuntimePrimitiveException(context, this,
          I18N.errorsJ().getN("org.nlogo.$common.thatAgentIsDead", turtle.classDisplayName()));
      }
      return turtle.getPatchHere().turtlesHere().iterator().hasNext();
    } else if (agentOrSet instanceof Patch) {
      return ((Patch) agentOrSet).turtlesHere().iterator().hasNext();
    } else if (agentOrSet instanceof AgentSet) {
      AgentSet sourceSet = (AgentSet) agentOrSet;
      if (sourceSet.kind() == AgentKindJ.Turtle()) {
        for (AgentIterator iter = sourceSet.iterator(); iter.hasNext();) {
          if (((Turtle) iter.next()).getPatchHere().turtlesHere().iterator().hasNext()) return true;
        }
        return false;
      } else {
        for (AgentIterator iter = sourceSet.iterator(); iter.hasNext();) {
          if (((Patch) iter.next()).turtlesHere().iterator().hasNext()) return true;
        }
        return false;
      }
    } else {
      throw new ArgumentTypeException
          (context, this, 0, Syntax.AgentType() | Syntax.AgentsetType(), agentOrSet);
    }
  }
}

File: netlogo-core/src/main/prim/_anyturtleson.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.{ AgentSet, Patch, Turtle }
import org.nlogo.core.{ AgentKind, I18N, Syntax }
import org.nlogo.nvm.{ ArgumentTypeException, Context, Reporter, RuntimePrimitiveException }

class _anyturtleson extends Reporter {
  override def report(context: Context): java.lang.Boolean =
    Boolean.box(report_1(context, args(0).report(context)))

  def report_1(context: Context, agentOrSet: AnyRef): Boolean = {
    agentOrSet match {
      case turtle: Turtle =>
        if(turtle.id == -1)
          throw new RuntimePrimitiveException(context, this, 
            I18N.errors.getN("org.nlogo.$common.thatAgentIsDead", turtle.classDisplayName))
        return true
      case patch: Patch =>
        val itr = patch.turtlesHere.iterator
        return itr.hasNext
      case sourceSet: AgentSet =>
        if(sourceSet.kind == AgentKind.Turtle) {
          val sourceSetItr = sourceSet.iterator
          return sourceSetItr.hasNext
        } else if(sourceSet.kind == AgentKind.Patch) {
          val sourceSetItr = sourceSet.iterator
          while (sourceSetItr.hasNext) {
            val patchItr = sourceSetItr.next().asInstanceOf[Patch].turtlesHere.iterator
            if(patchItr.hasNext)
              return true
          }
          false
        }
        false
      case _ =>
        throw new ArgumentTypeException(
          context, this, 0,
          Syntax.TurtleType | Syntax.PatchType |
          Syntax.TurtlesetType | Syntax.PatchsetType,
          agentOrSet)
    }
  }
}
